Add new `MachInst` backend and ARM64 support.

cfallin opened the pull request:
This PR contributes a new backend framework built around `MachInst`s (machine instructions) contained in `VCode` (virtual-registerized code) containers. Furthermore, it contains a working ARM64 backend built in this new framework. The new framework is properly connected into the main Cranelift compilation pipeline, and the appropriate integrations have been made so that `wasmtime` can run basic Wasm tests correctly on ARM64 hardware.

This patch sequence contains code written by Julian Seward <jseward@acm.org> and Benjamin Bouvier <public@benj.me>, originally developed on a side-branch before rebasing and condensing into this patch series. See the `arm64` branch at `https://github.com/cfallin/wasmtime` for original development history.
    
This patch also contains code written by Joey Gouly <joey.gouly@arm.com> and contributed to the above branch. These contributions are "Copyright (c) 2020, Arm Limited."

Closes #1174 and #1075.

sunfishcode commented on crates/runtime/src/traphandlers.rs:
```diff
@@ -70,6 +71,9 @@ cfg_if::cfg_if! {
             }
 
+            // on ARM64, we use `brk` to report traps, which generates SIGTRAP.
+            register(&mut PREV_SIGTRAP, libc::SIGTRAP);
```
Could you put this under a `cfg!(target_arch = "aarch64")` so that we don't register handlers for things we don't need to on x86?

cfallin replied:
Actually, a few days ago we found an alternative opcode that does deliver `SIGILL` on arm64, and I hadn't updated this yet -- so we can just remove it.

sunfishcode commented on tests/custom_signal_handler.rs:
```diff
@@ -121,8 +121,7 @@ mod tests {
             assert!(
-                trap.message()
-                    .starts_with("wasm trap: out of bounds memory access"),
+                trap.message().starts_with("wasm trap: out of bounds"),
```
This suggests you're generating an `OutOfBounds` trap code rather than a `HeapOutOfBounds` trap code. Is that a temporary limitation, or something more fundamental? If it's temporary, could you add a TODO comment here or so?

cfallin replied:
Hmm, I need to look into this one more. I remember making the edit thinking it was just a message change that a merge had missed, but I think you're correct that we're generating a different trap code now. We have some other failing tests on arm64 (i.e. only when running `cargo test` on arm64 hardware), so I'll revert and this just joins the club to be resolved later.

sunfishcode commented on crates/jit/src/link.rs:
```diff
@@ -132,6 +146,8 @@ cfg_if::cfg_if! {
             pub fn __rust_probestack();
         }
-        static PROBESTACK: unsafe extern "C" fn() = __rust_probestack;
+        static PROBESTACK: unsafe extern "C" fn() = empty_probestack;
```
Is this a temporary workaround? If so, could you add a comment here explaining what the real fix is? If not, could you add a comment explaining why probing isn't needed? :-)

cfallin replied:
Addressed -- it seems that stack probes are only supported on x86 / x86-64. I added a link to Rust's compiler-builtins impl of `__rust_probestack` where a comment describes the situation. Modified the code here to refer to `empty_probestack` only on non-x86, non-x86-64 architectures.

alexcrichton replied:
FWIW with https://github.com/bytecodealliance/wasmtime/pull/1490 we won't need this anymore anyway.

sunfishcode commented on cranelift/codegen/src/machinst/mod.rs:
```diff
@@ -0,0 +1,288 @@
+pub mod abi;
+pub use abi::*;
+pub mod pp;
+pub use pp::*;
```
`pp` isn't something I can guess the meaning of without more context. What would you think about renaming it to `pretty-print.rs` or so?

cfallin replied:
Renamed!

sunfishcode commented on cranelift/codegen/Cargo.toml:
```diff
@@ -33,7 +34,7 @@ byteorder = { version = "1.3.2", default-features = false }
 
 [features]
-default = ["std", "unwind"]
+default = ["std", "unwind", "all-arch"]
```
`all-arch` was deliberately removed from the default in https://github.com/bytecodealliance/cranelift/pull/853. Is there a new need for it to be enabled here, or is this something that could be enabled by downstream users of cranelift?

cfallin replied:
Sorry, that was added for development convenience only; reverted.

alexcrichton commented on crates/runtime/src/traphandlers.rs:
```diff
@@ -158,6 +158,12 @@ cfg_if::cfg_if! {
+                    extern "C" {
+                        fn GetPcFromUContext(cx: *mut libc::c_void) -> *const u8;
+                    }
+                    GetPcFromUContext(cx)
```
FWIW I'm adding the necessary Rust definitions here in https://github.com/rust-lang/libc/pull/1728, but this I think is the way to go for now.

bjorn3 commented on cranelift/codegen/src/machinst/mod.rs:
```diff
@@ -0,0 +1,288 @@
+    fn with_block_offsets(&mut self, my_offset: CodeOffset, targets: &[CodeOffset]);
+
+    /// Get the register universe for this backend.
+    fn reg_universe() -> RealRegUniverse;
```
I don't know how big this type is, but maybe return `&'static RealRegUniverse`?

bjorn3 replied:
Also doesn't it depend on which target features are enabled? (SSE, AVX, etc for x86)

cfallin replied:
In a subsequent commit on our dev branch, it becomes a function of the flags. I think this addresses both of your comments (for the former, we can't return a static instance because it might be dynamically computed per the settings).

bjorn3 commented on cranelift/codegen/src/machinst/mod.rs:
```diff
@@ -0,0 +1,288 @@
+    fn reg_universe(&self) -> RealRegUniverse;
+
+    /// Machine-specific condcode info needed by TargetIsa.
+    fn unsigned_add_overflow_condition(&self) -> IntCC {
```
Shouldn't this just be used during lowering?

cfallin replied:
Perhaps, but `TargetIsa` requires the information to come from somewhere. If we eventually remove more vestigial code when all backends migrate to the new framework, I think we'll be able to remove this altogether.

bjorn3 commented on cranelift/codegen/src/machinst/pp.rs:
```diff
@@ -0,0 +1,66 @@
+pub trait ShowWithRRU {
+    /// Return a string that shows the implementing object in context of the
+    /// given `RealRegUniverse`, if provided.
+    fn show_rru(&self, mb_rru: Option<&RealRegUniverse>) -> String;
```
```suggestion
    fn show_rru(&self, mb_rru: Option<&RealRegUniverse>, f: &mut std::fmt::Write);
```

This can avoid an allocation.

cfallin replied:
I was going to make this change at first, but it's quite a refactor to a bunch of debug-printing code (and more in the x64 backend which is not included here) which I'm not inclined to think is worth the effort. The `show_rru()` impls are meant to help debug-logging and are never called in production, so I think it's probably OK to keep this interface as-is, unless Dan/Ben/Julian disagree?

julian-seward1 replied:
Leave it as is, I say; this will never be used in production.

bjorn3 replied:
In cg_clif I write the clif ir before and after compilation to a file in debug mode. This makes debugging miscompilations much easier. It is unfortunately responsible for a nontrivial amount of compilation time, which means that I have to disable it in release mode. I would prefer if the new backend counter part to after compilation clif ir (vcode) could be printed as fast as possible.

cfallin replied:
That's a reasonable goal, but I suspect writing to the file will vastly outweigh allocation speeds here. In any case, the immediate costs are large -- a refactor, and a painful rebase of other work -- and the benefit is unclear and not driven by actual data. So let's leave this be for now, I think.

bjorn3 commented on cranelift/codegen/src/machinst/sections.rs:
```diff
@@ -0,0 +1,351 @@
+
+    /// Add 2 bytes to the section.
+    fn put2(&mut self, value: u16) {
+        self.put1((value & 0xff) as u8);
+        self.put1(((value >> 8) & 0xff) as u8);
```
Maybe use something like

```suggestion
        let [a, b] = value.to_be_bytes();
        self.put1(a);
        self.put1(b);
```

cfallin replied:
I learn something new about the standard library every day -- thanks!

julian-seward1 replied:
Is that really an improvement?  (1) It requires the reader to know (or look up) `to_be_bytes`; the original was quite clear.  (2) In order for it to not be a performance loss, it relies on `rustc` to inline away the call, the array construction and the array use, which is quite a big ask.

cfallin replied:
It seems that rustc actually does this; I just tested (with `-O`). Happy to do a little work to be idiomatic, as long as the complexity cost is not too high!

bjorn3 commented on cranelift/codegen/src/machinst/sections.rs:
```diff
@@ -0,0 +1,351 @@
+    /// Emit this section to the CodeSink and other associated sinks.  The
+    /// current offset of the CodeSink must match the starting offset of this
+    /// section.
+    pub fn emit<CS: CodeSink>(&self, sink: &mut CS) {
```
So the data is written to `MachSection` and then immediately copied to `CodeSink`? Why not write it to `CodeSink`?

cfallin replied:
The intent is to allow multiple sections to be constructed simultaneously; this rules out a streaming approach. (We've currently folded the constants and jumptables in as inline data on ARM64, but we previously actually build the sections this way, and could change again when we have proper support for long-range PC-relative addressing.)

bjorn3 commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -0,0 +1,738 @@
+    vreg_types: Vec<Type>,
+
+    /// Lowered machine instructions in order corresponding to the original IR.
+    pub insts: Vec<I>,
```
Use accessors instead? This type has several invariants that are easily broken by modifying `insts`.

cfallin replied:
Did one better and removed the `pub` qualifiers (only used for access by the `BlockRPO` computation).

bjorn3 commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -0,0 +1,738 @@
+        // block-start map given by the regalloc.
+        let block_ranges: Vec<(usize, usize)> =
+            block_ranges(result.target_map.elems(), result.insns.len());
+        let mut final_insns = vec![];
```
Can this be stored in `cranelift_codegen::Context` or some other context to prevent continuous reallocation.

cfallin replied:
I'm not sure what you mean by "continuous" -- this is one reallocation that happens per function compilation (`replace_insns_from_regalloc` is called once in the compilation pipeline). Holding a persistent vector to reuse in `Context` wouldn't really work with parallel compilation, if I understand your proposal correctly.

bjorn3 replied:
With continuous I mean that it reallocates for every function, instead of once per `Context`. (Apart reallocations necessary resize the vec.) The `Context` holds allocations that are reused between compilations of different functions. It works fine with parallel compilation, as you can just allocate a single `Context` per thread.

julian-seward1 replied:
As a general comment -- I have been profiling the new backend (Chris' stuff and the new regalloc) extensively with DHAT and plan to keep doing so.  That is very effective at finding pointless or low-value repeated reallocation -- in the week 6-13 Apr I reduced the number of allocation calls by a factor of 2 to 3, and there is much more to come.  Given that such allocation-avoiding restructurings potentially make code harder to understand, I'd prefer to restrict them to cases where the profile numbers show there's a clear benefit to be had, or that the change is very local and so doesn't impact readability -- eg, replacing a `Vec` with a `SmallVec`.

bjorn3 commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -0,0 +1,738 @@
+}
+
+// N.B.: Debug impl assumes that VCode has already been through all compilation
+// passes, and so has a final block order and offsets.
```
This means that you can't just `println!("{:?}", vcode);` before all compilation passes for debugging.

cfallin replied:
The comment here is wrong, on closer inspection -- removed. (If `final_block_order` is empty, that line of the output just shows an empty list.)

bjorn3 commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -0,0 +1,738 @@
+        // as final_block_order, but also includes blocks which are in the
+        // representation but not in final_block_order.
+        let mut display_order = Vec::<usize>::new();
+        // First display blocks in |final_block_order|
```
```suggestion
        // First display blocks in `final_block_order`
```

bjorn3 commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -0,0 +1,738 @@
+    }
+}
+
+// Pretty-printing with `RealRegUniverse` context.
```
```suggestion
/// Pretty-printing with `RealRegUniverse` context.
```

Rustdoc supports doc comments on `impl`'s

bjorn3 commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -0,0 +1,738 @@
+        }
+
+        let mut s = String::new();
+        s = s + &format!("VCode_ShowWithRRU {{{{");
```
```suggestion
        s.push_str("VCode_ShowWithRRU {{");
```

bjorn3 replied:
Why is this printing two `{` by the way.

cfallin replied:
This is @julian-seward1's debug-printing format; I'm not entirely sure but I am sure there is a good reason for it :-) Perhaps it allows for easier log-grepping. I'll resolve this for now but we can always come back and refine the debug-printing format later if it needs to be more concise, or machine-readable, or whatever.

julian-seward1 replied:
Just because I like these top level debug markers to be easy to spot in the thousands of lines of debug output.  No other reason.

bjorn3 commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -0,0 +1,738 @@
+
+        let mut s = String::new();
+        s = s + &format!("VCode_ShowWithRRU {{{{");
+        s = s + &"\n".to_string();
```
```suggestion
        s.push('\n');
```

bjorn3 commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -0,0 +1,738 @@
+                    "** OMITTED **"
+                } else {
+                    ""
+                })
+                .to_string();
```
```suggestion
                });
```

bjorn3 commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -0,0 +1,738 @@
+            s = s + &format!("  (instruction range: {} .. {})", start, end);
+            s = s + &"\n".to_string();
+            for inst in start..end {
+                s = s + &format!(
```
```suggestion
                write!(s,
```

You may need to import `std::fmt::Write` though.

bjorn3 commented on cranelift/codegen/src/isa/arm64/inst/regs.rs:
```diff
@@ -0,0 +1,273 @@
+    if reg.is_real() {
+        // Change (eg) "x42" into "w42" as appropriate
+        if reg.get_class() == RegClass::I64 && is32 && s.starts_with("x") {
+            s = "w".to_string() + &s[1..];
```
```suggestion
            s.replace_range(0..1, "w");
```

bjorn3 commented on cranelift/codegen/src/isa/arm64/inst/regs.rs:
```diff
@@ -0,0 +1,273 @@
+    } else {
+        // Add a "w" suffix to RegClass::I64 vregs used in a 32-bit role
+        if reg.get_class() == RegClass::I64 && is32 {
+            s = s + &"w";
```
```suggestion
            s.push('w');
```

bjorn3 commented on cranelift/codegen/src/isa/arm64/inst/regs.rs:
```diff
@@ -0,0 +1,273 @@
+        return s;
+    }
+    let prefix = if is32 { "s" } else { "d" };
+    prefix.to_string() + &s[1..]
```
```suggestion
    s.replace_range(0..1, prefix);
    s
```

bjorn3 commented on cranelift/codegen/src/isa/arm64/inst/regs.rs:
```diff
@@ -0,0 +1,273 @@
+    if reg.is_real() {
+        // Change (eg) "v0" into "d0".
+        if reg.get_class() == RegClass::V128 && s.starts_with("v") {
+            s = "d".to_string() + &s[1..];
```
Same as above

bjorn3 commented on cranelift/codegen/src/isa/arm64/inst/args.rs:
```diff
@@ -0,0 +1,501 @@
+
+impl ShiftOp {
+    /// Get the encoding of this shift op.
+    pub fn bits(&self) -> u8 {
```
This can take `self` by value.

bjorn3 commented on cranelift/codegen/src/isa/arm64/inst/args.rs:
```diff
@@ -0,0 +1,501 @@
+    }
+
+    /// Return the shift amount.
+    pub fn value(&self) -> u8 {
```
```suggestion
    pub fn value(self) -> u8 {
```

bjorn3 commented on cranelift/codegen/src/isa/arm64/lower.rs:
```diff
@@ -0,0 +1,2805 @@
+            let mem = lower_address(ctx, elem_ty, &inputs[1..], off);
+            let rd = input_to_reg(ctx, inputs[0], NarrowValueMode::None);
+
+            let memflags = ctx.memflags(insn).expect("memory flags");
```
The `aligned` memflag is never checked. Are stores always lowered to unaligned stores, or will it just emit a signal on unaligned stores?

cfallin replied:
Apparently, aarch64 supports unaligned accesses with the ordinary load/store instructions. I just tested this on real hardware and it seems to work; [1] also confirms.

[1] https://stackoverflow.com/questions/38535738/does-aarch64-support-unaligned-access

bjorn3 commented on cranelift/codegen/src/isa/arm64/mod.rs:
```diff
@@ -1,6 +1,219 @@
+}
+
+/// Create a new `isa::Builder`.
+pub fn isa_builder(triple: Triple) -> IsaBuilder {
```
Assert that the triple has `aarch64` as isa?

bjorn3 commented on cranelift/codegen/src/isa/arm64/mod.rs:
```diff
@@ -1,6 +1,219 @@
+    IsaBuilder {
+        triple,
+        setup: settings::builder(),
+        constructor: isa_constructor,
```
```suggestion
        constructor: |_, shared_flags, _| {
            let backend = Arm64Backend::new_with_flags(shared_flags);
            Box::new(TargetIsaAdapter::new(backend))
        },
```

If a closure doesn't capture any variables, it can coerce to a function pointer.

bjorn3 commented on cranelift/codegen/src/context.rs:
```diff
@@ -149,23 +172,33 @@ impl Context {
-        debug!("Compiled:\n{}", self.func.display(isa));
-        result
+        if let Some(backend) = isa.get_mach_backend() {
+            let func = std::mem::replace(&mut self.func, Function::new());
```
Previously you could just reuse the allocations inside `Function, but this will discard all allocations inside `Function`.

bjorn3 commented on cranelift/codegen/src/context.rs:
```diff
@@ -275,13 +312,19 @@ impl Context {
+            // Legalization invalidates the domtree and loop_analysis by mutating the CFG.
+            // TODO: Avoid doing this when legalization doesn't actually mutate the CFG.
+            self.domtree.clear();
+            self.loop_analysis.clear();
```
Why are these clears no longer done for the new backend?

bjorn3 commented on cranelift/codegen/src/context.rs:
```diff
@@ -275,13 +312,19 @@ impl Context {
+            self.loop_analysis.clear();
+            legalize_function(&mut self.func, &mut self.cfg, isa);
+            debug!("Legalized:\n{}", self.func.display(isa));
+            self.verify_if(isa)
```
Please keep this for the new backend.

bjorn3 commented on cranelift/codegen/src/ir/function.rs:
```diff
@@ -17,6 +19,7 @@ use crate::isa::{CallConv, EncInfo, Encoding, Legalize, TargetIsa};
 use crate::regalloc::{EntryRegDiversions, RegDiversions};
 use crate::value_label::ValueLabelsRanges;
 use crate::write::write_function;
+use alloc::boxed::Box;
```
This doesn't seem to be used.

bjorn3 commented on cranelift/codegen/src/isa/test_utils.rs:
```diff
@@ -0,0 +1,83 @@
+
+    /// This is pretty lame, but whatever ..
+    pub fn stringify(&self) -> String {
+        let mut s = "".to_string();
+        for b in &self.bytes {
+            s = s + &format!("{:02X}", b).to_string();
+        }
+        s
```
```suggestion
        use std::fmt::Write;
        let mut s = String::new();
        for b in &self.bytes {
            write!(s, "{:02X}, b);
        }
        s
```

bjorn3 commented on cranelift/filetests/filetests/vcode/arm64/arithmetic.clif:
```diff
@@ -0,0 +1,242 @@
+test vcode arch=arm64
```
Why not use the existing `target aarch64` directive?

cfallin replied:
The simple answer is that I couldn't work out how to fetch it. It doesn't appear to be included in the `preamble_comments` of the `Context`. It doesn't seem too important to me; but I'm happy to change it if you can point out the proper plumbing to make use of here.

cfallin replied:
Actually, upon second look, worked this out -- the tests now use the `target` directive. Resolving!

bjorn3 commented on crates/jit/src/link.rs:
```diff
@@ -142,12 +142,17 @@ cfg_if::cfg_if! {
+        // https://github.com/rust-lang/compiler-builtins/blob/cae3e6ea23739166504f9f9fb50ec070097979d4/src/probestack.rs#L39,
+        // LLVM only has stack-probe support on x86-64 and x86. Thus, on any other CPU
+        // architecture, we simply use an empty stack-probe function.
+        extern "C" fn empty_probestack() {}
```
The `probestack` has an ABI that is not compatible with `C`. This means that this will only work correctly if the compiler happened to decide that only a `ret` should be emitted. If it does anything else, like writing to a caller saved register that was used as argument for the caller of `probestack`, you will likely get a crash.

cfallin replied:
Yes, that's a theoretical concern. In practice, the empty function `{}` compiles to just a `ret` in both debug and release mode on both x86-64 and aarch64 (just tested all four combinations). Do you have a suggestion for a better approach (compiler attributes, etc)?

bjorn3 replied:
Maybe use Cranelift to create the function?

cfallin replied:
That seems like quite a bit of overhead, and potentially fragile and complex code to maintain (a new "probestack" ABI to implement on ARM64, a codegen step at startup, a way to lazily initialize the `PROBESTACK` global, etc.) for a theoretical concern that does not actually occur in practice. I'll defer to @sunfishcode 's judgment here, but my inclination is to leave it as-is.

sunfishcode replied:
As it turns out, @alexcrichton's https://github.com/bytecodealliance/wasmtime/pull/1490 will eliminate the need for this call entirely, which avoids this question entirely.

Beyond that, if we want to reintroduce stack probes, we should add a `__rust_probestack` for AArch64 and any other platform we need it on [here](https://github.com/rust-lang/compiler-builtins/blob/master/src/probestack.rs) rather than having our own version in Cranelift.

cfallin replied:
@sunfishcode, that sounds perfect. Would you like us to wait for #1490 to land, or would it be acceptable to land this with the temporary "empty function" stack probe in the meantime (if the timing works out that way)?

sunfishcode replied:
I think it's fine to land it with this temporary workaround.

bnjbvr commented on cranelift/codegen/src/machinst/compile.rs:
```diff
@@ -0,0 +1,76 @@
+    debug!("vcode from lowering: \n{}", vcode.show_rru(Some(universe)));
+
+    // Perform register allocation.
+    let algorithm = match env::var("REGALLOC") {
```
Actually, I think it'd be even better to defer this choice to later; LSRA isn't properly working yet (I'll switch back when i'm done with the Spidermonkey integration), so we shouldn't give a false sentiment of safety to people by allowing them to use it. So I'd strongly recommend to just always select the backtracking allocator.

bjorn3 replied:
Maybe name the LSRA option `experimental_lsra`?

cfallin replied:
Simply reverted to `RegAllocAlgorithm::Backtracking` for now; we can keep an env var backdoor as part of a local development-setup patch if needed, but for now the only stable/fast choice for mainline is BT, without checker, I think (Ben/Julian: correct me if wrong, if you think we should enable checker by default).

bjorn3 commented on cranelift/codegen/src/isa/arm64/mod.rs:
```diff
@@ -1,132 +1,218 @@
-    fn regclass_for_abi_type(&self, ty: ir::Type) -> RegClass {
-        abi::regclass_for_abi_type(ty)
+    fn triple(&self) -> Triple {
+        FromStr::from_str("arm64").unwrap()
```
This should be the triple passed to `isa_builder`. This value will result in a binary format of unknown, which means that `cranelift_object` will not work.

bjorn3 replied:
Also `target-lexicon` calls it `aarch64`.

cfallin replied:
I agree, we might as well standardize on the official name. Renamed the backend to `aarch64`. There are still a few vestigial remnants of `Arm64` around, e.g. the `Arm64Call` reloc type which is exposed in the public API.

sunfishcode commented on cranelift/codegen/src/ir/function.rs:
```diff
@@ -3,6 +3,8 @@
 //! The `Function` struct defined in this module owns all of its basic blocks and
 //! instructions.
 
+#![allow(unused_imports)]
```
This may be an artifact of development, but we should avoid adding warning overrides like this in the main tree unless there's a practical reason we can't just fix the warning.

sunfishcode commented on cranelift/codegen/src/postopt.rs:
```diff
@@ -360,20 +360,23 @@ fn optimize_complex_addresses(pos: &mut EncCursor, inst: Inst, isa: &dyn TargetI
-                {
-                    if constraints.clobbers_flags {
-                        last_flags_clobber = Some(inst)
+                if !is_mach_backend {
```
The optimization here is converting a comparison opcode that produces an integer result and a branch that consumes an integer result into a comparison that produces a flags result and a branch that consumes the flags result. Currently it does this by rewriting the comparison in place, so there could be instructions between the compare and the branch. If they clobber the flags, the optimization isn't correct. So it isn't safe to disable this check. A possible alternative though would be to change the optimization to move the comparison just before the branch, to ensure that the flags value isn't clobbered.

cfallin replied:
Ah, OK -- the arm64 backend at least will correctly handle an `icmp` + `brz`/`brnz` sequence, so we don't need this optimization -- I've just altered it to skip this mutation entirely when a MachBackend is in use.

sunfishcode commented on cranelift/codegen/src/verifier/flags.rs:
```diff
@@ -28,13 +28,17 @@ pub fn verify_flags(
-        livein: SecondaryMap::new(),
-    };
-    verifier.check(errors)
+    if isa.is_none() || isa.unwrap().get_mach_backend().is_none() {
```
Can you comment on why it makes sense to disable the FlagsVerifier with the mach backend?

cfallin replied:
The FlagsVerifier as written uses the encoding info to determine whether instructions clobber the flags. We could certainly build an equivalent with the new infrastructure, or else add a target-independent notion of "clobbers flags" to the CLIF opcodes themselves, but it would be some extra plumbing. What do you think?

sunfishcode replied:
The flags verifier can still check target-independent properties -- the `encinfo` is an `Option` to allow it to run without this information, so I think we can still enable the verifier, just by passing `None` for the `encinfo` argument when the mach backend is enabled.

cfallin replied:
Ah, yes, that makes sense -- done!

cfallin commented on cranelift/codegen/src/machinst/abi.rs:
```diff
@@ -0,0 +1,142 @@
+    fn gen_ret(&self) -> I;
+
+    /// Generate an epilogue placeholder.
+    fn gen_epilogue_placeholder(&self) -> I;
```
Added comment.

cfallin commented on cranelift/codegen/src/machinst/abi.rs:
```diff
@@ -0,0 +1,142 @@
+
+/// Trait implemented by an object that tracks ABI-related state (e.g., stack
+/// layout) and can generate code while emitting the *body* of a function.
+pub trait ABIBody<I: VCodeInst> {
```
Done, thanks!

cfallin commented on cranelift/codegen/src/machinst/abi.rs:
```diff
@@ -0,0 +1,142 @@
+    /// frame or other setup necessary to use the other methods (`load_arg`,
+    /// `store_retval`, and spillslot accesses.)  |self| is mutable so that we
+    /// can store information in it which will be useful when creating the
+    /// epilogue.
```
Eh, I think that's probably unnecessary complexity -- the `ABIBody` is specialized to, and only lives for the duration of, one function compilation anyway.

cfallin commented on cranelift/codegen/src/machinst/abi.rs:
```diff
@@ -0,0 +1,142 @@
+
+    /// Returns the full frame size for the given function, after prologue emission has run. This
+    /// comprises the spill space, incoming argument space, alignment padding, etc.
+    fn frame_size(&self) -> u32;
```
Also unnecessary complexity IMHO; we'd then need to store it twice to allow the outer layer (VCode) to return information upward.

cfallin commented on cranelift/codegen/src/machinst/compile.rs:
```diff
@@ -0,0 +1,76 @@
+        vcode.show_rru(Some(universe))
+    );
+
+    //println!("{}\n", vcode.show_rru(Some(&B::MInst::reg_universe())));
```
Done.

cfallin commented on cranelift/codegen/src/machinst/lower.rs:
```diff
@@ -0,0 +1,723 @@
+            | &InstructionData::LoadComplex { flags, .. }
+            | &InstructionData::Store { flags, .. }
+            | &InstructionData::StoreComplex { flags, .. } => Some(flags),
+            _ => None,
```
I'm not sure it's reasonable to list out all `InstructionData` variants here and in the other helpers, as there are (at present) 50 of them; that's quite a lot of verbosity. Ideally we would add new accessors to the generated code. I'm not as familiar with the DSL mechanisms for that; thoughts?

julian-seward1 replied:
I would just leave it as it is.  CL will fail at run time the first time someone implementing a new instruction format tries to run any test case, so it'll be obvious it needs to be fixed.

cfallin commented on cranelift/codegen/src/machinst/lower.rs:
```diff
@@ -0,0 +1,723 @@
+        for bb in bbs.iter() {
+            debug!("lowering bb: {}", bb);
+
+            // If this is a return block, produce the return value setup.
```
Instructions are lowered in reverse order, so I think this is actually correct (albeit counterintuitive!). I'll add a comment to explain this better.

cfallin commented on cranelift/codegen/src/machinst/compile.rs:
```diff
@@ -0,0 +1,76 @@
+    let result = {
+        let _tt = timing::regalloc();
+        allocate_registers(
+            &mut vcode, algorithm, universe, /*request_block_annotations=*/ false,
```
OK, I haven't used rust-analyzer before; but for basic source-code readability, I'd prefer to keep the argument-name comment here.

cfallin commented on cranelift/codegen/src/machinst/blockorder.rs:
```diff
@@ -0,0 +1,59 @@
+        let (start, end) = &vcode.block_ranges[block as usize];
+        for i in *start..*end {
+            if vcode.insts[i as usize].is_epilogue_placeholder() {
+                debug_assert!(self.deferred_last.is_none());
```
Yes, but this only happens when the "fallthrough return" mode is used (grep for `GenerateReturn::No` in `machinst/lower.rs`, `gen_retval_setup()`), which comes from SpiderMonkey (the Baldrdash ABI). In that case, a return in the middle of the function doesn't make sense anyway.

cfallin commented:
Thanks @sunfishcode and @bjorn3 for the comments -- I think I've addressed everything, except for the open questions in the unresolved comments above.

bjorn3 commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -0,0 +1,879 @@
+    (int_saves, vec_saves)
+}
+
+fn is_caller_save(call_conv: isa::CallConv, r: RealReg) -> bool {
```
And check for call conv here too.

cfallin replied:
Added check to toplevel `new()`.

bjorn3 commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -0,0 +1,879 @@
+
+    fn liveins(&self) -> Set<RealReg> {
+        let mut set: Set<RealReg> = Set::empty();
+        for arg in &self.sig.args {
```
If arg is `Copy` you can do:
```suggestion
        for &arg in &self.sig.args {
```

bjorn3 commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -0,0 +1,879 @@
+    fn load_stackslot(
+        &self,
+        slot: StackSlot,
+        offset: usize,
```
Please take `u64` or `i64` instead. This is platform dependent, so won't work nice for cross-compilation.

cfallin replied:
Will use `u32`, actually; we don't expect a stackframe to be larger than 4GB.

bjorn3 commented on cranelift/codegen/src/isa/aarch64/inst/regs.rs:
```diff
@@ -0,0 +1,274 @@
+    // X29
+    61,
+    // X30
+    62,
```
Who came up with this order? It is even worse than `eax`, `ecx`, `edx`, `ebx` on x86.

cfallin replied:
I am glad we could provide some entertainment :-) It's actually intentional: the register allocator library (regalloc.rs) expects all allocatable registers in a contiguous index range. Because the ABI and codegen constraints poke random holes in this range, the mapping has to be a piecewise-linear thing. It's ugly, but it works...

bjorn3 commented on cranelift/codegen/src/isa/aarch64/lower.rs:
```diff
@@ -0,0 +1,2815 @@
+                    let imm: i64 = imm.into();
+                    Some(imm as u64)
+                }
+                &InstructionData::UnaryIeee32 { opcode: _, imm } => Some(imm.bits() as u64),
```
```suggestion
                &InstructionData::UnaryIeee32 { opcode: _, imm } => Some(u64::from(imm.bits())),
```

bjorn3 commented on cranelift/codegen/src/machinst/lower.rs:
```diff
@@ -0,0 +1,727 @@
+    }
+}
+
+fn branch_targets(f: &Function, block: Block, inst: Inst) -> SmallVec<[Block; 16]> {
```
Also I think 16 elements is more than likely will be used most of the cases.

cfallin replied:
Done by reworking as `visit_branch_targets()` (internal iteration, accepts a closure, monomorphized at each call site).

bjorn3 commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -0,0 +1,733 @@
+        // Compute block offsets.
+        let mut code_section = MachSectionSize::new(0);
+        let mut block_offsets = vec![0; self.num_blocks()];
+        for block in &self.final_block_order {
```
```suggestion
        for &block in &self.final_block_order {
```

bjorn3 commented on cranelift/codegen/src/machinst/vcode.rs:
```diff
@@ -0,0 +1,733 @@
+            let new_offset = I::align_basic_block(code_section.cur_offset_from_start());
+            while new_offset > code_section.cur_offset_from_start() {
+                // Pad with NOPs up to the aligned block offset.
+                let nop = I::gen_nop((new_offset - code_section.cur_offset_from_start()) as usize);
```
What if nops are always say 4 bytes, but there needs to be a 3 byte offset to align to the required 4 bytes. That would mean `I::gen_nop` will always need to return a 0 byte nop, which means that the loop will never terminate.

cfallin replied:
The invariant that the machine backend must satisfy is that `gen_nop` returns an instruction with nonzero size, and must be smaller than the requested size, so by that invariant, this loop will always terminate. Here's the doc comment from `gen_nop()` for reference:

    /// Generate a NOP. The `preferred_size` parameter allows the caller to                                                                     
    /// request a NOP of that size, or as close to it as possible. The machine                                                                  
    /// backend may return a NOP whose binary encoding is smaller than the                                                                      
    /// preferred size, but must not return a NOP that is larger. However,                                                                      
    /// the instruction must have a nonzero size.

sunfishcode commented on cranelift/codegen/src/isa/aarch64/inst/mod.rs:
```diff
@@ -0,0 +1,2515 @@
+//! This module defines aarch64-specific machine instruction types.
+
+#![allow(non_snake_case)]
+#![allow(unused_imports)]
```
Could you remove these few remaining `unused_imports` directives in a few files?

cfallin replied:
Done!

sunfishcode commented on cranelift/codegen/src/isa/aarch64/inst/mod.rs:
```diff
@@ -0,0 +1,2515 @@
+
+#![allow(non_snake_case)]
+#![allow(unused_imports)]
+#![allow(non_camel_case_types)]
```
Several files have `allow(non_snake_case)` and `allow(non_camel_case_types)`. Could you either remove these and change the names to follow the regular Rust conventions, or add comments explaining why exceptions to the rules are justified in these files?

cfallin replied:
Removed! None of the names were actually unconventional; I think this was left over from an earlier prototype.

sunfishcode commented on cranelift/codegen/src/isa/aarch64/inst/emit.rs:
```diff
@@ -0,0 +1,4106 @@
+        Some(ExtendOp::UXTW) => 0b010,
+        Some(ExtendOp::SXTW) => 0b110,
+        Some(ExtendOp::SXTX) => 0b111,
+        None => 0b011, /* LSL */
```
Minor nit: use a `//`-style comment here.

sunfishcode commented on cranelift/codegen/src/isa/aarch64/inst/emit.rs:
```diff
@@ -0,0 +1,4106 @@
+                sink.put4(enc_fcsel(rd, rn, rm, cond, /* is32 = */ true));
+            }
+            &Inst::FpuCSel64 { rd, rn, rm, cond } => {
+                sink.put4(enc_fcsel(rd, rn, rm, cond, /* is32 = */ false));
```
There are a lot of `is32` parameters being passed around in a lot of code. Would it be worth adding a dedicated enum type to use in place of bool, to catch errors and make the code more self-documenting?

cfallin replied:
Good idea -- done!

cfallin commented on cranelift/codegen/src/context.rs:
```diff
@@ -279,9 +315,15 @@ impl Context {
-        self.verify_if(isa)
+        if isa.get_mach_backend().is_some() {
+            // Run some specific legalizations only.
+            simple_legalize(&mut self.func, &mut self.cfg, isa);
```
The new backend takes a different approach. We want to get away from the DSL-based pattern/legalization/recipe world and write the backends in a more explicit way. This should lead to more simplicity and maintainability (one of the main motivations of the new backend design). E.g., in the ARM64 backend, we handle different operand widths explicitly. It still makes sense to do some expansions with the existing code, such as for `heap_addr` or `global_value`, which is where `simple_legalize()` comes in.

bjorn3 replied:
Many legalizations are shared between backends. Also legalizations may enable optimizations that would otherwise have to be rewritten for each backend.

cfallin replied:
OK -- we can incorporate legalizations as required, then, into the new pipeline, I think. The issue with the original request that I see is that "when there is no encoding" is no longer meaningful: we are no longer taking the approach of directly mapping CLIF instructions to hardware encodings and editing in place otherwise. Pulling in some of the existing legalizations here is meant to be an intermediate state, in place until we transition fully to the new backend and then can rewrite the expansions as a more straightforward pass. @sunfishcode and I have talked about various ways of having a "pattern-matching" approach to this.

cfallin commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -0,0 +1,879 @@
+    // Arguments only: on stack, at given offset from SP at entry.
+    Stack(i64, ir::Type),
+    // (first and only) return value only: in memory pointed to by x8 on entry.
+    #[allow(dead_code)]
```
Yeah, actually, probably better to remove the partial impl of `RetMem` until we actually need it. Thanks.

cfallin commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -0,0 +1,879 @@
+/// This is SpiderMonkey's `WasmTlsReg`.
+static BALDRDASH_TLS_REG: u8 = 23;
+
+// These two lists represent the registers the JIT may *not* use at any point in generated code.
```
@bnjbvr would be the person to answer this (this is his SpiderMonkey-integration code); Ben, comments?

cfallin replied:
Resolving this for now, as Ben is on top of ABI issues and has done further work on caller/callee-saves that will come in another PR.

cfallin commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -0,0 +1,879 @@
+    }
+}
+
+fn is_callee_save(call_conv: isa::CallConv, r: RealReg) -> bool {
```
Done; added check to toplevel `new()` instead.

cfallin commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -0,0 +1,879 @@
+        // sizeof(ty).
+        let islot = slot.get() as i64;
+        let ty_size = self.get_spillslot_size(into_reg.to_reg().get_class(), ty) * 8;
+        let fp_off: i64 = -(self.stackslots_size as i64) - (8 * islot) - ty_size as i64;
```
Any particular reason (and if so, why not throughout the entire codebase for every `as` cast)?

bjorn3 replied:
This prevents accidential truncation of integers.

cfallin replied:
Ah! I had been under the impression that it was a type error to cast to narrower integers. I'll go back and audit these now.

cfallin replied:
(But in this particular case, we're casting from `u32` to `i64`, so truncation should be an issue.)

sunfishcode replied:
`as` casts silently pick between widening, narrowing, and value-changing casts, while `::from` casts get a type error on any potentially value-changing cast, so it's nice to use `::from` when applicable.

cfallin commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -0,0 +1,879 @@
+
+        if !self.call_conv.extends_baldrdash() {
+            // The MOV (alias of ORR) interprets x31 as XZR, so use an ADD here.
+            // MOV to SP is an alias of ADD.
```
Just added to `emit.rs`, thanks.

cfallin commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -0,0 +1,879 @@
+    }
+
+    fn frame_size(&self) -> u32 {
+        self.frame_size
+            .expect("frame size not computed before prologue generation") as u32
```
Using `u32` in stack-frame layout code instead; as above, we don't expect >4GB frames.

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/args.rs:
```diff
@@ -0,0 +1,500 @@
+    pub fn reg_maybe_offset(reg: Reg, offset: i64, value_type: Type) -> Option<MemArg> {
+        if offset == 0 {
+            Some(MemArg::Unscaled(reg, SImm9::zero()))
+        } else if let Some(simm9) = SImm9::maybe_from_i64(offset) {
```
Ah, yes, this was left over from an earlier version that used a different mode; removed redundant case.

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/args.rs:
```diff
@@ -0,0 +1,500 @@
+    pub fn lower(&mut self, targets: &[CodeOffset], my_offset: CodeOffset) {
+        match self {
+            &mut BranchTarget::Block(bix) => {
+                let bix = bix as usize;
```
Eh, perhaps it's just a style thing, but I don't like these casts that require trait imports and this `unwrap()`; it complexifies the code for no good reason, and makes it harder to audit the code for potentially panic'ing operations.

In fact, the `bix as usize` does one better if we do ever try to build on a 16-bit host: it will statically fail to compile, because `u32 as u16` is not a valid cast. I'd prefer that over an `unwrap()` that fires at runtime.

bjorn3 replied:
`u32 as u16` is actually valid. It truncates.

cfallin replied:
Yup, just realized this -- thanks!

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/args.rs:
```diff
@@ -0,0 +1,500 @@
+    pub fn map(&mut self, block_index_map: &[BlockIndex]) {
+        match self {
+            &mut BranchTarget::Block(ref mut bix) => {
+                let n = block_index_map[*bix as usize];
```
Not making this change for same reason as above; would prefer a static compilation failure to a dynamic panic.

cfallin replied:
(Actually making this change now, as per realization above.)

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/emit.rs:
```diff
@@ -0,0 +1,4106 @@
+                    | ALUOp::MSub64
+                    | ALUOp::SMulH
+                    | ALUOp::UMulH => {
+                        //// RRRR ops.
```
No, this one is intentional; the RRRR ops are the 3-source, 1-dest ops.

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/imms.rs:
```diff
@@ -0,0 +1,753 @@
+        assert!(scale_ty == I64 || scale_ty == I32);
+        let scale = scale_ty.bytes();
+        assert!(scale.is_power_of_two());
+        let scale = scale as i64;
```
Holding off on this for now, as above. I'm curious if there's an idiomatic-Rust reason for this -- is there some downside to using the `x as T` operator built into the language?

cfallin replied:
Updated now; thanks for the education on Rust casts!

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/imms.rs:
```diff
@@ -0,0 +1,753 @@
+pub struct MoveWideConst {
+    /// The value.
+    pub bits: u16,
+    /// shifted 16*shift bits to the left.
```
Ambiguous subject, I suppose; the result is `bits` shifted to the left. Edited to clarify.

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/mod.rs:
```diff
@@ -0,0 +1,2515 @@
+    And64,
+    AndNot32,
+    AndNot64,
+    Eor32,
```
Yep, it's an ARM-ism. In general we tried to maintain consistent naming w.r.t. the instruction manual unless there was a good reason otherwise. Added a doc comment to clarify.

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/mod.rs:
```diff
@@ -0,0 +1,2515 @@
+    MAdd64,
+    MSub32,
+    MSub64,
+    SMulH,
```
High word result of multiply (e.g., bits 127-64 for the i64 case), either signed or unsigned variant. Added doc comment.

cfallin commented on cranelift/codegen/src/isa/aarch64/inst/mod.rs:
```diff
@@ -0,0 +1,2515 @@
+        match ty {
+            I8 | I16 | I32 | I64 | B1 | B8 | B16 | B32 | B64 => RegClass::I64,
+            F32 | F64 => RegClass::V128,
+            I128 | B128 => RegClass::V128,
```
Removed. I'm not sure if NEON has a plain 128-bit load or not, but we plan to handle vector ops after our initial merge and evaluation, so I'll defer the proper implementation of this to later.

cfallin commented on cranelift/codegen/src/isa/aarch64/lower.rs:
```diff
@@ -0,0 +1,2815 @@
+        // that register.
+        if narrow_mode != NarrowValueMode::None
+            && ((narrow_mode.is_32bit() && out_bits < 32)
+                || (!narrow_mode.is_32bit() && out_bits < 64))
```
Unreachable because of surrounding conditional's `narrow_mode != NarrowValueMode::None`.

cfallin commented on cranelift/codegen/src/machinst/lower.rs:
```diff
@@ -0,0 +1,727 @@
+/// A context that machine-specific lowering code can use to emit lowered instructions. This is the
+/// view of the machine-independent per-function lowering context that is seen by the machine
+/// backend.
+pub trait LowerCtx {
```
The intention is to also eventually have other `Lower` impls. For one, there has been some conversation about verification efforts that plug in a different impl to get symbolic lowerings out (this is very handwavy now, but we want to leave the door open). For another, we might eventually want to drive the lowering logic with a different IR, and so we're trying to keep as much behind this API as possible (it's not perfect, e.g. `Opcode` type, but this will help with that goal). Everything is monomorphized, so the only cost now should be a little extra type verbosity.

cfallin commented on cranelift/codegen/src/machinst/lower.rs:
```diff
@@ -0,0 +1,727 @@
+        self.vcode.push(inst);
+    }
+
+    fn find_reachable_bbs(&self) -> SmallVec<[Block; 16]> {
```
We didn't want to rely on the domtree here, for the same reason as above (portability); though perhaps a followup PR could plumb the domtree info through the `LowerCtx`.

cfallin commented on cranelift/codegen/src/machinst/lower.rs:
```diff
@@ -0,0 +1,727 @@
+    }
+
+    fn find_reachable_bbs(&self) -> SmallVec<[Block; 16]> {
+        if let Some(entry) = self.f.layout.entry_block() {
```
It might trigger errors elsewhere, but I had wanted to provide a robust impl here; the code should still do the right thing (i.e., nothing) when presented with an empty function body.

bjorn3 replied:
The code would fallthrough to the next function when there are no blocks, which shouldn't be allowed.

cfallin replied:
Yes, perhaps, if the verifier allows this; but it seems that the old backend would have the same problem, no? (There would be no instruction to have an encoding, thus no return would be emitted.) It seems this would be good to follow up in a separate PR if it is indeed allowed, by updating the verifier instead. (As a general code-style thing I don't want to distribute implicit assumptions about data structure state throughout the code; if the entry is an `Option<Block>` then let's write code everywhere that supports that, and if it should always be a `Some`, then let's just make it a `Block` instead.)

cfallin commented on cranelift/codegen/src/machinst/lower.rs:
```diff
@@ -0,0 +1,727 @@
+
+        // Allocate a separate BlockIndex for each control-flow instruction so that we can create
+        // the edge blocks later. Each entry for a control-flow inst is the edge block; the list
+        // has (cf-inst, edge block, orig block) tuples.
```
Control-flow instruction; expanded.

cfallin commented on cranelift/codegen/src/machinst/lower.rs:
```diff
@@ -0,0 +1,727 @@
+        let mut bbs = self.find_reachable_bbs();
+        // Work backward (reverse block order, reverse through each block), skipping insns with zero
+        // uses.
+        bbs.reverse();
```
At first I had thought this wouldn't work, because we initialize the vcode BB map with `bbs` just below this; but it turns out that that code is invariant to the block order (we'll just get different `BlockIndex`es in the vcode, but this is arbitrary anyway). So, done!

cfallin commented on cranelift/codegen/src/machinst/lower.rs:
```diff
@@ -0,0 +1,723 @@
+                    }
+
+                    // Only codegen an instruction if it either has a side
+                    // effect, or has at least one use of one of its results.
```
Nope, not required; the reasoning behind including it here is that it is intertwined with the operand-matching/merging behavior. For example, an `iadd` can incorporate an `iconst` on an input; it then calls `ctx.merged(input)` which indicates that the use was incorporated directly, updating the refcounts (use counts). This could just as well be handled by a later inst-level DCE, but it seemed cheaper to avoid the lowering logic in the first place (Julian's measurement earlier today was ~4K host x86 instructions spent to lower one target instruction; that should improve but still). If you think we should factor these apart, I'm happy to do that, though!

cfallin commented on cranelift/codegen/src/machinst/lower.rs:
```diff
@@ -0,0 +1,723 @@
+        let mut bbs = self.find_reachable_bbs();
+        // Work backward (reverse block order, reverse through each block), skipping insns with zero
+        // uses.
+        bbs.reverse();
```
It's actually a BFS (`find_reachable_bbs()` further up in this file), but the order doesn't matter here, as we *also* compute RPO later in blockorder.rs :-)

My thinking here was that I wanted to avoid creating a dependency on the `domtree` in order to remain as portable as possible, i.e., given later alternative implementations of the frontend driving the lowering logic, and that the `O(blocks)` cost would hopefully not be too bad. (I reserve the right to change my mind later in the week when I start looking at profiling results in detail!) But I can plumb this through if you'd prefer.

sunfishcode replied:
Makes sense!

cfallin commented:
Thanks @sunfishcode and @bjorn3 for the further review comments! I think I've addressed everything except for the open questions in unresolved threads above. Happy to take further review comments, of course.

bjorn3 commented on cranelift/codegen/src/isa/aarch64/inst/args.rs:
```diff
@@ -0,0 +1,538 @@
+//! AArch64 ISA definitions: instruction arguments.
+
+// Some variants are never constructed, but we still want them as options in the future.
+#[allow(dead_code)]
```
This applies only to the `use crate::binemit::CodeOffset;`.

bjorn3 commented on cranelift/codegen/src/isa/aarch64/inst/args.rs:
```diff
@@ -0,0 +1,538 @@
+            Cond::Gt => 12,
+            Cond::Le => 13,
+            Cond::Al => 14,
+            Cond::Nv => 15,
```
You could use the same `#[repr(u8)` + `Eq = 0` + `self as u8` here as `ShiftOp` and `ExtendOp`.

bjorn3 commented on cranelift/codegen/src/isa/aarch64/inst/args.rs:
```diff
@@ -0,0 +1,538 @@
+
+/// Type used to communicate whether an instruction is a 32-bit variant or not. This `is32` boolean
+/// flag occurs often, so we make it an enum for safety.
+#[derive(Clone, Copy, Debug)]
+pub enum Is32 {
+    Yes,
+    No,
+}
```
```suggestion
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Size {
    Size32,
    Size64,
}
```

`Eq` can be used as replacement for `is32` and `is64` if you think that isn't worse.

If you have a better suggestion for the name of the variants, feel free to use that instead.

bjorn3 commented on cranelift/codegen/src/isa/aarch64/inst/args.rs:
```diff
@@ -0,0 +1,538 @@
+    }
+}
+
+impl From<bool> for Is32 {
+    fn from(val: bool) -> Is32 {
+        if val {
+            Is32::Yes
+        } else {
+            Is32::No
+        }
+    }
+}
```
This should be removed if you accept the suggestion above.

bjorn3 commented on cranelift/codegen/src/machinst/adapter.rs:
```diff
@@ -0,0 +1,126 @@
+
+impl fmt::Display for TargetIsaAdapter {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "MachBackend")
```
```suggestion
        f.debug_struct("MachBackend")
            .field("name", self.backend.name())
            .field("triple", self.backend.triple())
            .field("flags", self.backend.flags())
            .finish()
```

bjorn3 commented on cranelift/codegen/src/verifier/flags.rs:
```diff
@@ -28,10 +28,15 @@ pub fn verify_flags(
     errors: &mut VerifierErrors,
 ) -> VerifierStepResult<()> {
     let _tt = timing::verify_flags();
+    let encinfo = if isa.is_none() || !isa.unwrap().get_mach_backend().is_none() {
```
```suggestion
    let encinfo = if isa.is_none() || isa.unwrap().get_mach_backend().is_some() {
```

bjorn3 commented on cranelift/codegen/src/isa/test_utils.rs:
```diff
@@ -0,0 +1,87 @@
+        TestCodeSink { bytes: vec![] }
+    }
+
+    /// This is pretty lame, but whatever ..
```
```suggestion
    // This is pretty lame, but whatever ..
```

bjorn3 commented on cranelift/codegen/src/isa/aarch64/inst/regs.rs:
```diff
@@ -0,0 +1,270 @@
+    }
+}
+
+/// If |ireg| denotes an I64-classed reg, make a best-effort attempt to show
```
```suggestion
/// If `ireg` denotes an I64-classed reg, make a best-effort attempt to show
```

bjorn3 commented on cranelift/codegen/src/isa/aarch64/lower.rs:
```diff
@@ -0,0 +1,2777 @@
+                Opcode::Call => {
+                    let extname = ctx.call_target(insn).unwrap();
+                    let extname = extname.clone();
+                    // HACK: get the function address with an Abs8 reloc in the constant pool.
+                    //let tmp = ctx.tmp(RegClass::I64, I64);
+                    //ctx.emit(Inst::LoadExtName {
+                    //rd: tmp,
+                    //name: extname,
+                    //srcloc: loc,
+                    //offset: 0,
+                    //});
```
Please remove this.

cfallin commented:
Alright, I think I've addressed all comments, finally. Let's wait to see if the latest CI run succeeds; then, @sunfishcode, if you're OK with it, I can hit the Merge button (or feel free to do so yourself if that's the custom here). Thanks for all the review work!

bjorn3 commented on cranelift/codegen/src/machinst/compile.rs:
```diff
@@ -0,0 +1,63 @@
+/// Compile the given function down to VCode with allocated registers, ready
+/// for binary emission.
+pub fn compile<B: LowerBackend>(
+    f: &mut Function,
```
This can take &Function I think.

cfallin replied:
Yup, this was left over from an earlier version that did mutate the function; much nicer now to take an immutable input. Thanks!

bjorn3 commented on cranelift/codegen/src/isa/aarch64/mod.rs:
```diff
@@ -0,0 +1,220 @@
+        AArch64Backend { triple, flags }
+    }
+
+    fn compile_vcode(&self, func: &mut Function, flags: &settings::Flags) -> VCode<inst::Inst> {
```
Same here

bjorn3 commented on cranelift/codegen/src/isa/aarch64/mod.rs:
```diff
@@ -0,0 +1,220 @@
+impl MachBackend for AArch64Backend {
+    fn compile_function(
+        &self,
+        func: &mut Function,
```
Same here

cfallin commented:
@sunfishcode actually, I don't have permissions to merge, so please feel do hit the big green button when you're happy!

cfallin commented:
(Actually, I take that back, Till just added me. Let me know if you're happy with the final version and I'll merge.)

cfallin commented:
As per IM with @sunfishcode, merging now!

peterhuene commented on cranelift/codegen/src/isa/mod.rs:
```diff
@@ -48,16 +48,17 @@ pub use crate::isa::call_conv::CallConv;
-use crate::isa::enc_tables::Encodings;
-#[cfg(feature = "unwind")]
 use crate::isa::fde::RegisterMappingError;
+#[cfg(feature = "unwind")]
```
Unfortunately I just caught this.  I  believe this accidentally moved the `#[cfg(feature = "unwind")]`.

I have a PR that this merged conflicted with, so I'll correct it there.

cfallin replied:
Yikes, that was likely a rebase failure on my part. Sorry about that. (I suppose I got lucky as `unwind` is enabled by default.) Thanks for fixing!
